First draft of ipush/pop

christian-herber-nxp opened the pull request.

christian-herber-nxp committed:
Smip First draft of ipush/pop

tovine reviewed (commented):
Very good proposal! We should also try to get reviews from someone in the Zephyr community and other RTOS'es

christian-herber-nxp commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,149 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+* Saves hart's state of the interrupted context that can be overwritten in case of preemption
+* Save registers x10-x15 (these are the caller saved registers available in RVC)
+* Adjusts the stack pointer to create the stack frame
+* Optionally, enables interrupts
```
yes, it will be clear once i put the encoding in there

christian-herber-nxp commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,149 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+* Disables interrupts
+* Restores registers x10-x15
+* Restores the the hart's state to the point that the hart can return to the interrupted context
+* Adjust the stack pointer to destroy the stack frame
```
not sure. If this is the case, then a ipop instruction is needed in addition to ipopxret.

tovine replied:
Couldn't this be added as a "property" of the instruction if such a crypto extension is present/enabled?
Similarly to how Zkt doesn't add any new instructions but rather just adds a requirement for the affected instructions (if implemented) to execute in constant time: https://github.com/rvkrypto/riscv-zkt-list/blob/main/zkt-list.adoc

christian-herber-nxp replied:
could be. for this discussion, i would leave this as out of scope, to limit the things we need to deal with.
honestly, i think it would be legal for an implementation to zero out a destroyed stack frame (which could quite easily done with cache block operations (in case you have a dcache).

tovine replied:
Good point, this would apply to other instructions as well and belong in that potential future crypto extension instead of this one. :+1:

christian-herber-nxp commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,149 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+
+#TODO: Discuss if we want to have the illegal instruction exceptions as defined here
+
+With these requirements, implementations are allowed to speculatively start executing an ipush instruction in case of a taken interrupt.
```
yes, but still, a change would be observable that is not expected. Needs to be decided if acceptable.

christian-herber-nxp commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,149 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+
+NOTE: In the typical case of tail chaining, the check for another pending enabled interrupt is done after disabling interrupts at the start of the ipopxret instruction. In this case, only the conditional interrupt enablement of the ipush instruction would need to executed.
+
+Once the instruction retires, or the interrupts are disabled, no late preemption can occur.
```
the scenario is not fully clear to me. During the time in which the chained interrupt handler is being fetched, i think there is another window for late preemption

christian-herber-nxp commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,149 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+
+According the the Machine Level ISA, trap handlers must be designed to not cause exceptions during their critical section.
+This applies to the ipush and ipopxret instruction.
+Therefore, by design, it must be guaranteed that no faults occur during the execution of these instructions.
```
this is the note from the privileged ISA

> Trap handlers must be designed to neither enable interrupts nor cause exceptions during
> the phase of handling where the trap handler preserves the critical state information
> required to handle and resume from the trap. An exception or interrupt in this critical
> phase of trap handling may lead to a trap that can overwrite such critical state. This could
> result in the loss of data needed to recover from the initial trap. Further, if an exception
> occurs in the code path needed to handle traps, then such a situation may lead to an
> infinite loop of traps. To prevent this, trap handlers must be meticulously designed to
> identify and safely manage exceptions within their operational flow.

I think this means you stack must be in a place where no exceptions can occur as a result of an access.

christian-herber-nxp commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,149 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+* It is better to harmonize this, and to consistently save/restore the interrupted context
+** This would fix the issues discovered earlier where mcause in the trampoline is overwritted before reading it in the handler
+* For this, while not necessary with the ipush/ipopret proposals, we might consider to introduce pepc and xpexccode fields. For the ipush/ipopret, these names are just used to refer to the values of the interrupted context
+* the necessary changes would also change the behavior of xret, where now mcause.exccode is being overwritten with xpexccode upon execution of xret.
```
the difference between pcause and cause is that pcause is independent of the interrupt, but just dependend on the interrupted context. Thus, easier to tail-chain, late-preempt etc.

mark-honman commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,149 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+The interrupt push and pop extension adds instructions to accelerate interrupt handling, which additionally reduce code size.
+Interupt latency and jitter is reduced by allowing the context save to happen speculatively in parallel to the vector table and handler fetch.
+Additional latency reduction is achieved by allowing late-preemption of the push sequence in favor of a higher priority interrupt.
+Interrupt throughput is increased by allowing tail-chaing of interupts, skipping the context restore and save in between servicing interrupts.
```
There may not be a big latency saving for late preemption, because the vector table and handler fetch for the preempting interrupt cannot be overlapped with register value push activity.

christian-herber-nxp committed:
Smip: Adding clearing of the ip bit to the push sequence

christian-herber-nxp committed:
Smcspsw: Adding proposal for a conditional scratch swap instruction

christian-herber-nxp committed:
Smip: Adding conditional scratch swap to ipush/ipop sequences

christian-herber-nxp committed:
Smcspsw: Use dedicated xsp registers over xscratch

christian-herber-nxp committed:
Smcspsw: Changing to single register for stack pointer swapping
- You do not need to hold two stack pointers at a given point in time
- if you leave a given privilege mode to a lower mode, you are only coming back via a trap
- thus there is no need to have to stacks active at a single mode after starting the lower mode

tovine commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,199 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+
+stack pointer adjustment: 32
+
+32: reserved (5) | xpp (2) |  xpie (1) |  xpil (8) | reserved (4) | xpexccode (12)
```
This looks almost like `xcause` but not entirely (no `interrupt` bit, and some fields are shuffled slightly) - is this the proposed layout of a new `xpcause` CSR or similar? If so, it would be better to refer to it by this name here, and then specify the register layout and the individual bit fields somewhere else in the spec.

christian-herber-nxp replied:
the layout is just a strawman. I just want something in place and we can decide later on what the best layout would be.

tovine commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,199 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+
+=== Debug
+
+#TODO: Define what happens if there is a break point at the ipush instruction address
```
If possible, I think we should define this in a way that doesn't prevent speculative state saving before the handler address has been fetched (since we need the table lookup to know if it's a break point hit) as that could add many precious cycles to the interrupt latency.

christian-herber-nxp replied:
agree. ipush cannot be burdened. 
i do not think there is a good solution to it. i think it will end up in a way that a breakpoint on the ipush instruction may be able to observe a partial update or even complete update of ipush. 
another route would be to not allow breakpoints on ipush, or to say a matching breakpoint will break on the subsequent instruction.

tovine replied:
Can we add that restriction to breakpoints? In that case it sounds like a reasonable solution. I could see some pushback, but it's worth trying

christian-herber-nxp replied:
something is in now and ready to be challenged/improved

tovine reviewed (approved):
Changes look good, added some new comments below but nothing critical.

christian-herber-nxp commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,199 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+== Smip Interrupt push and pop extension
+
+The interrupt push and pop extension adds instructions to accelerate interrupt handling, which additionally reduce code size.
+Interupt latency and jitter is reduced by allowing the context save to happen speculatively in parallel to the vector table and handler fetch.
```
iirc, it is a requirement in Zcmp that push restarts from the beginning if interrupted.
We need to define something similar here.

christian-herber-nxp replied:
i did recall correctly:

> If a trap occurs during the sequence, then any loads which were executed before the trap may update
> architectural state. The loads will be re-executed once the trap handler completes, so the values will be
> overwritten. Therefore it is permitted for an implementation to update some of the destination
> registers before taking a fault.

Really, this is not an issue that wouldn't exist without ipush/ipop

tovine replied:
Very good to see that we thought about that, then it shouldn't be an issue! (sometimes it's a bit hard to remember)

christian-herber-nxp commented on src/clic.adoc:
```diff
@@ -2053,6 +2053,199 @@ the setting in the CLIC Configuration register (`mcliccfg.xnlbits`).
+
+== Smip Interrupt push and pop extension
+
+The interrupt push and pop extension adds instructions to accelerate interrupt handling, which additionally reduce code size.
```
tend to disagree with the suggestions. the short version of my sentence is "The interrupt push and pop extension adds instructions, which additionally reduce code size."

which seems fine.

christian-herber-nxp committed:
Smip: Fixed typo

Co-authored-by: Torbjørn Viem Ness <tovine@users.noreply.github.com>
Signed-off-by: Christian Herber <christian.herber@oss.nxp.com>

christian-herber-nxp committed:
Smip: Added notes on expectations if the ipush/ipopxret instructions are interrupted

christian-herber-nxp committed:
Smip: Fixed heading level of sections in push/pop section

christian-herber-nxp committed:
Smip: Defined behavior for ipush for breakpoints

christian-herber-nxp committed:
Smcspsw: Fixed the logic for stack pointer swap

christian-herber-nxp committed:
Smcspsw: Included exceptions in the logic for stack pointer swap

christian-herber-nxp committed:
Smip: Added the interrupt flag to the saved context

christian-herber-nxp committed:
Smcspsw: Removed xscratchcs(l) CSRs and replaced with xsp

christian-herber-nxp committed:
Smcspsw: Updated handler examples with xcsw instruction

christian-herber-nxp committed:
Smic: reword the ip bit clearing

christian-herber-nxp committed:
Smip: Add handling of double trap bit to push/pop

christian-herber-nxp committed:
Requiring xstatus.exccode to be placed in a0 during push

christian-herber-nxp commented:
changes were incorporated in a different PR. closing this here.
